xapian-core/cluster/kmeans.cc

```diff
@@ -0,0 +1,139 @@
+/** @file kmeans.cc
+ *  @brief KMeans clustering API
+ */
+/* Copyright (C) 2016 Richhiey Thomas
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian/cluster.h"
+#include <debuglog.h>
+#include <xapian/error.h>
+#include <omassert.h>
+
+#include <climits>
+#include <random>
+#include <set>
+#include <unordered_map>
+#include <vector>
+
+// Threshold value for checking convergence in KMeans
+#define EPSILON 0.0000000001
+
+using namespace Xapian;
+using namespace std;
+
+KMeans::KMeans(unsigned int k_, unsigned int max_iters_)
+: k(k_), max_iters(max_iters_) {
```

Comment:
Xapian convention is to indent the member initialisers by one level (so four spaces before this `:`).  There are quite a few instances of this below too.